src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs

```diff
@@ -204,11 +201,12 @@ internal static SemaphoreSlim GetThrottleSemaphoreSlim(IEnvironmentVariableReade
             }
 
             cancellationToken.ThrowIfCancellationRequested();
-            Task<LibraryIdentity> result = FindLibraryIdentityAsync(libraryRange, cacheContext);
 
             try
             {
-                return await result;
+                LibraryIdentity result = await _libraryMatchCache.GetOrAddAsync(libraryRange, cacheContext.RefreshMemoryCache, () => FindLibraryCoreAsync(libraryRange, cacheContext, logger, cancellationToken), cancellationToken);
```

Comment:
Isn't this a change from the previous implementation where we were calling it using Task.Run?

Reply:
I'm certainly no expert in TPL for these advance cases but I thought calling `MethodAsync()` and `Task.Run(() => MethodAsync())` is essentially the same thing.  So there's no need to do so since this is calling a method which receives the task factory delegate which actually schedules the task and returns a task that can be awaited and shared by callers to all await.  In all of my testing, things seem to work?

Reply:
I don't have a comment on the change, however:

> but I thought calling MethodAsync() and Task.Run(() => MethodAsync()) is essentially the same thing

No, that is not the same thing. 

Given the following method:

``` C#
Task MethodAsync() 
{
    File.Delete("Foo.cs");
    
    await FooAsync();
}
```


``` C#
await MethodAsync();
```

``` C#
await Task.Run(() => MethodAsync());
```

`await MethodAsync();` will synchronously call File.Delete on the same thread, yielding the calling thread (releasing back to the pool) at `await FooAsync`.

`await Task.Run(() => MethodAsync());` will schedule a new item onto the thread pool where File.Delete will execute when there is a free thread, yielding the calling thread immediately.

Reply:
Ah so is it safe to say that `Task.Run()` should always be used in case the called method has a mixture of async/sync code?

Reply:
No, in fact it can [hurt your throughput](https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki/24027/Threading?anchor=%E2%9D%8C-**avoid**-unneeded-thread-pool-queuing-via-%60task.run%60).  Calling Task.Run and immediately awaiting on it will only hurt (you saying "add this task to the back of the queue").

You want to use Task.Run when scheduling multiple mostly synchronous/CPU tasks to enable them to overlap and run at the same time.

Reply:
I think in this situation, it probably doesn't make a difference whether we have an extra Task.Run, since we're in an async codepath and are always going to be on the threadpool already.

Reply:
@Erarndt The Task.Run was added recently here in https://github.com/NuGet/NuGet.Client/pull/5556.

Reply:
@davkean If we want to avoid blocking the caller in an async path, is it preferrable to do `Task.Run()` or use `Task.Yield()`?

Reply:
In this exact case (running on a thread pool thread), they are mostly identical. If you are running on something different than a thread pool, they will be different (Task.Yield on a UI thread for example, would get readded to back of message queue),